proxy_agent/src/shared_state.rs

```diff
@@ -0,0 +1,222 @@
+        } else {
+            state.wireserver_rule_id = rule_id;
+            (true, old_rule_id)
+        }
+    }
+
+    pub fn get_wireserver_rule_id(shared_state: Arc<Mutex<SharedState>>) -> String {
+        shared_state.lock().unwrap().wireserver_rule_id.to_string()
+    }
+
+    /// Update the IMDS rule ID
+    /// # Arguments
+    /// * `shared_state` - Arc<Mutex<SharedState>>
+    /// * `rule_id` - String
+    /// # Returns
+    /// * `bool` - true if the rule ID is update successfully
+    /// * `String` - the rule Id before the update operation
+    pub fn update_imds_rule_id(
+        shared_state: Arc<Mutex<SharedState>>,
+        rule_id: String,
+    ) -> (bool, String) {
+        let mut state = shared_state.lock().unwrap();
+        let old_rule_id = state.imds_rule_id.clone();
+        if old_rule_id == rule_id {
+            (false, old_rule_id)
+        } else {
+            state.imds_rule_id = rule_id;
+            (true, old_rule_id)
+        }
+    }
+
+    pub fn get_imds_rule_id(shared_state: Arc<Mutex<SharedState>>) -> String {
+        shared_state.lock().unwrap().imds_rule_id.to_string()
+    }
+
+    pub fn set_shutdown(shared_state: Arc<Mutex<SharedState>>, shutdown: bool) {
+        shared_state.lock().unwrap().key_keeper_shutdown = shutdown;
+    }
+
+    pub fn get_shutdown(shared_state: Arc<Mutex<SharedState>>) -> bool {
+        shared_state.lock().unwrap().key_keeper_shutdown
+    }
+
+    pub fn set_status_message(shared_state: Arc<Mutex<SharedState>>, status_message: String) {
+        shared_state.lock().unwrap().key_keeper_status_message = status_message;
+    }
+
+    pub fn get_status_message(shared_state: Arc<Mutex<SharedState>>) -> String {
+        shared_state
+            .lock()
+            .unwrap()
+            .key_keeper_status_message
+            .to_string()
+    }
+}
+
+pub mod proxy_listener_wrapper {
+    use super::SharedState;
+    use std::sync::{Arc, Mutex};
+
+    pub fn set_shutdown(shared_state: Arc<Mutex<SharedState>>, shutdown: bool) {
+        shared_state.lock().unwrap().proxy_listner_shutdown = shutdown;
+    }
+
+    pub fn get_shutdown(shared_state: Arc<Mutex<SharedState>>) -> bool {
+        shared_state.lock().unwrap().proxy_listner_shutdown
+    }
+
+    /// Increase the connection count
+    /// # Arguments
+    /// * `shared_state` - Arc<Mutex<SharedState>>
+    /// # Returns
+    /// * `u128` - the updated connection count
+    /// # Remarks
+    /// * If the connection count reaches u128::MAX, it will reset to 0
+    pub fn increase_connection_count(shared_state: Arc<Mutex<SharedState>>) -> u128 {
+        let mut state = shared_state.lock().unwrap();
+        let mut connection_count = state.connection_count;
+        if connection_count == u128::MAX {
+            // reset connection count
+            connection_count = 0;
+        }
+        connection_count += 1;
+        state.connection_count = connection_count;
+        connection_count
```

Comment:
FYI there's [overflowing_add](https://doc.rust-lang.org/std/primitive.u128.html#method.overflowing_add) on primitive integer types to express you want that behavior, so you could write this as 
```suggestion
        (state.connection_count, _) = state.connection_count.overflowing_add(1);
        state.connection_count
```